v3

cometkim opened the pull request:
https://github.com/cometkim/gatsby-plugin-typegen/milestone/3

Todos
-------
- [x] Predictable & Debuggable scheduler (Thanks to XState)
- [x] Remove graphql-tools dependency and integrate more inside Gatsby
- [x] Integration with stable lifecycle API (`SET_SCHEMA` and `SET_GRAPHQL_DEFINITIONS`)
- Clean implementation & Tests
  - [x] codegen
  - [x] emitSchema
  - [x] emitPluginDocument
  - [x] autofix
- [x] E2E Test on example project
  - [x] Add TypeScript example
  - [x] Add Flow example

Issues
--------

- [x] Resolves #40 
- [x] Resolves #103 
- [x] Resolves #104 
- [x] #109 
- [x] Fixes #113 
- [x] Fixes #117 
- [x] Fixes #120 
- [x] Resolves #122 
- [x] Fixes #124 
- [x] Resolves #131 
- [x] Fixes #139, #99 
- [x] Resolves #146 
- [x] Fixes #157 
- [x] Fixes #159
- [x] Fixes #160 
- [x] Fixes #161
- [x] Fixes #162
- [x] Resolves #168 
- [x] Resolves #169 
- [x] Fixes #171

cometkim committed:
prepare v3 release

cometkim committed:
BREAKING CHANGE(plugin): fix graphql-js peer version to v14

Gatsby uses GraphQL 14 and the graphql-tools is compatible with both v14 and v15.

Fixes #120
See also https://github.com/gatsbyjs/gatsby/issues/25906

cometkim committed:
chore: use upstream's type definitions

cometkim committed:
WIP: implemente XState based scheduler

cometkim committed:
refactor: autoFixService

cometkim committed:
impl scheduler and test in an example project

cometkim committed:
chore: fix outputPath default

cometkim committed:
WIP

cometkim committed:
lexicographic sort schema

cometkim committed:
stabilize schema output

cometkim committed:
bump GCG

cometkim commented:
I took a long break from working on this project.

I still have a lot of ideas and things to improve, but I'm losing focus. (even though I'm still actively using this plugin for my gatsby projects)

Maybe we need to get another maintainer after this work. please contact me if anyone has an interest.

cometkim committed:
upgrade yarn to v3

cometkim committed:
update and cleanup root dependencies

cometkim committed:
rewritten core logics

cometkim committed:
upgrade eslint to v8

cometkim committed:
upgrade plugin's dependencies

cometkim committed:
pin graphql version

cometkim committed:
remove unnecessary

cometkim committed:
update sdks

cometkim committed:
disable P'n'P

cometkim committed:
fix tests

cometkim committed:
chore

cometkim committed:
rewrite emitSchema service

cometkim committed:
rewriten emitPluginDocument service

cometkim committed:
rewriten codegen service

cometkim committed:
integrate all

cometkim committed:
fix and cleanup all

cometkim committed:
remove old example

cometkim committed:
fix all

cometkim committed:
add TS & Flow examples

cometkim commented:
This has been a burden on my mind for a long time.
Now I think most of the work is done and I've fixed almost all the issues.

The v3 changes are mostly minor changes to make internal behavior stable.

There are several deprecated options, but with a simple name change, the operation is maintained up to v4, so you can update the version without worrying about the site being broken.

Here is some highlight.

- gatsby-plugin-typegen is now **very fast**, **very stable** This comes from the integration of stable Gatsby events with a predictable scheduler built with XState.
- There was a major refactoring. Getting a readable and testable codebase will be a must for the next contributors & maybe a new maintainer.
- `emitPluginDocuments` option has been renamed to `emitPluginDocument`. The previous option will remain only in v3.
- `autoFix` option has been renamed to `autofix`. Keep the old option only in v3.  The previous option will remain only in v3.
- Added a basic TypeScript example (to ensure it works)
- Added a basic Flow example (to ensure it works)
- Got rid of many bugs (20 issues)

I really appreciate [Easy Agile](https://twitter.com/easyagile) team for continuing to fund the project while it was being silent. Your funds will be used for future contributions and maintenance.

cometkim commented:
gatsby-plugin-typegen@v3.0.0-rc.0 is availiable

cometkim commented:
I will fix the CI and some documentation before release.

In the meantime, try it on your project and send me feedback

LekoArts commented:
I'd assume you could move the yarn changes to another PR or already apply the yarn/`.gitignore` changes to the main branch? Then this PR will have way less changed files and the changes made are clearer :)

cometkim commented:
I found a bug in the Flow codegen, but fortunately I don't think it's fatal, as Flow doesn't break the full type checking and still maintains high type coverage via powerful inference.

The issue can be fixed upstream eventually.

LekoArts commented on plugin/src/gatsby-node.ts:
```diff
@@ -1,257 +1,203 @@
+      .default('src/__generated__/gatsby-types.d.ts'),
+    includeResolvers: Joi.boolean()
+      .default(false),
+    autoFix: Joi.boolean()
```
You'll want to remove this old key, Gatsby will automatically warn about invalid keys

When I tried out the RC I got `warn [typegen] The 'autoFix' option is deprecated. Use 'autofix' instead.` so I think it gets set due to this here

LekoArts commented:
This PR seems to remove a lot of files in `plugin/.yarn/cache`, can you also do this in master and rebase then? The GitHub UI is _really_ laggy with this amount of files 😆

LekoArts commented:
Notes from trying out the RC (I'll add more as I try it out):

- When there are no fragments the `gatsby-plugin-documents.graphql` doesn't get generated. Should it be generated but have some placeholder?
- When trying the VSCode extension the "Execute query" only works if the above mentioned `gatsby-plugin-documents.graphql` contains something, if it's empty/not existent it doesn't work
- When starting the server where a query has a name, then removing the name, and re-adding the name, the type for the query doesn't exist anymore unless the server is restarted

cometkim commented:
There hasn't been an update since I've been busy with business lately, but it would be all checked and released this week or early next week.

I will test the compatibility of the new version of Gatsby with Gatsby Cloud in [my public production](https://github.com/daangn/websites).

cometkim commented:
@LekoArts VSCode extension's feature should work with valid graphql-config, or if invalid, it must be reported by it.

The `emitPluginDocument` option exists to extract the [pre-defined fragments](https://github.com/gatsbyjs/gatsby/blob/22b9c2f/packages/gatsby-transformer-sharp/src/fragments.js) by the plugin or gatsby core and should not have any effect on the project if not used.

It would be a good idea to keep valid settings together in examples.

cometkim commented:
The plugin intentionally creates types only for named queries.

It requires using a globally unique query name for maintaining the correct colocation pattern and is also a limitation of both Gatsby and GraphQL Codegen.

When it is not given, Gatsby uses the pathname to make it internally, but this is inappropriate as a type name and can be easily broken by moving files.

cometkim commented:
> This PR seems to remove a lot of files in plugin/.yarn/cache, can you also do this in master and rebase then? The GitHub UI is really laggy with this amount of files laughing

@LekoArts I've tried it, but it seems that GitHub has already given up tracking changes here :sweat_smile: 

The code has been rewritten almost from scratch, so it will be better to walk around the code directly in [the branch](https://github.com/cometkim/gatsby-plugin-typegen/tree/v3) than to read the changes.

cometkim commented:
I've checked that this works well in production and doesn't cause any issues on Gatsby Cloud (Fixes #109)

However, it seems that some documents included within the theme are not generated properly.

And `flattenGeneratedTypes` option makes generated types a bit bloated

kvchen commented:
Is it possible to add the `flattenGeneratedTypesIncludeFragments: true` option in addition to `flattenGeneratedTypes`? In v2, the exported types included named fragments - that seems to have gone away in v3. This makes it difficult to use typed GraphQL fragments in subcomponents.

I'd file a new pull request, but the change is so small that it might be easier to just add it to this pull request.

cometkim committed:
rc.0

cometkim committed:
skip running on cloud build

cometkim committed:
fix config validation

cometkim committed:
more contexture reporting

cometkim committed:
turn flattenGeneratedTypesIncludeFragments true

cometkim committed:
stabilize documents

cometkim committed:
fix some bugs

cometkim committed:
cleanup code a bit

cometkim committed:
sync flow result

cometkim committed:
change maybe type to use null instead of undefined

cometkim committed:
fix test

cometkim committed:
update README

cometkim committed:
rc.1

cometkim committed:
add mdx example

MDX example to validate regression of [#117]

[#117]: https://github.com/cometkim/gatsby-plugin-typegen/issues/117

cometkim committed:
update dependencies

cometkim committed:
added support for `GatsbyImageData` scalar

cometkim committed:
rc.2

cometkim commented:
Merged to [main](https://github.com/cometkim/gatsby-plugin-typegen/tree/main)
